### PR TITLE
Create data factory middleware

### PR DESCRIPTION
**Pull request by thewilkybarkid**
Moves the data factory from the context prototype to a dedicated middleware with its own context, which will allow for more use of the context in future middleware.

Extracted from https://github.com/libero/article-store/pull/115.

### CODE REVIEW

**Review by giorgiosironi — APPROVED** [review]
The types don't seem like the best bang for the buck, but the refactoring is clear

**Comment by thewilkybarkid** [comment]
Upped the typing game in https://github.com/libero/article-store/pull/115/commits/59ad69abcce8cf711ece32ebe0211199c9702528, will move that here.

File: src/middleware/data-factory.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import {
+  DefaultContext, DefaultState, Middleware, Next,
+} from 'koa';
+import { DataFactory } from 'rdf-js';
+
+export type DataFactoryContext<CustomT extends object = DefaultContext,
+  CustomF extends object = {}> = CustomT & { dataFactory: DataFactory & CustomF };
+
+export default <CustomF extends object>(dataFactory: DataFactory & CustomF):
```
Comment (thewilkybarkid) [inline src/middleware/data-factory.ts]: Trying to change the generic to `<CustomF extends DataFactory>`, but running in to a few problems in `data-factory.test.ts`.

File: src/app.ts [hunk 2]
```diff
@@ -6,17 +6,17 @@ import logger from 'koa-logger';
...
+import setDataFactory, { DataFactoryContext } from './middleware/data-factory';
 import emptyResponse from './middleware/empty-response';
 import errorHandler from './middleware/error-handler';
 import jsonld from './middleware/jsonld';
 import routing from './middleware/routing';
 
 export type AppState = DefaultState;
 
-export type AppContext = RouterContext<AppState, {
+export type AppContext = RouterContext<AppState, DataFactoryContext<{
```
Comment (giorgiosironi) [inline src/app.ts:17]: where's the type for `RouterContext`? DefinitelyTyped or somewhere else?

File: src/middleware/data-factory.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import {
+  DefaultContext, DefaultState, Middleware, Next,
+} from 'koa';
+import { DataFactory } from 'rdf-js';
+
+export type DataFactoryContext<CustomT extends object = DefaultContext,
+  CustomF extends object = {}> = CustomT & { dataFactory: DataFactory & CustomF };
```
Comment (giorgiosironi) [inline src/middleware/data-factory.ts]: `T` and `F` don't sound very explanatory, they stand for `Type` and?

File: src/app.ts (hunk 2 above)
Comment (thewilkybarkid) [reply src/app.ts:17]: Yep https://github.com/libero/article-store/blob/dd041f613487e7b39f72ba3c3a4f3fcc8e89e3da/src/app.ts#L2

File: src/middleware/data-factory.ts (hunk 3 above)
Comment (thewilkybarkid) [reply src/middleware/data-factory.ts]: `CustomT` using same name as Koa. `F` for factory, but these could be better.

File: src/app.ts (hunk 2 above)
Comment (giorgiosironi) [reply src/app.ts:17]: ah, I didn't see it in https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types because the list was too long

**Pull request merged**